egs/librispeech/asr/nnlm/local/model.py

```diff
@@ -0,0 +1,185 @@
+# reference:
+# https://github.com/pytorch/examples/blob/master/word_language_model/model.py
+import math
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class RNNModel(nn.Module):
+    """Container module with an encoder, a recurrent module, and a decoder."""
+
+    def __init__(self,
+                 rnn_type,
+                 ntoken,
+                 ninp,
+                 nhid,
+                 nlayers,
+                 dropout=0.5,
+                 tie_weights=False):
+        super(RNNModel, self).__init__()
+        self.ntoken = ntoken
+        self.drop = nn.Dropout(dropout)
+        # import pdb; pdb.set_trace()
+        self.encoder = nn.Embedding(ntoken, ninp, padding_idx=0)
+        if rnn_type in ['LSTM', 'GRU']:
+            self.rnn = getattr(nn, rnn_type)(ninp,
+                                             nhid,
+                                             nlayers,
+                                             dropout=dropout)
+        else:
+            try:
+                nonlinearity = {
+                    'RNN_TANH': 'tanh',
+                    'RNN_RELU': 'relu'
+                }[rnn_type]
+            except KeyError:
+                raise ValueError(
+                    """An invalid option for `--model` was supplied,
+                                 options are ['LSTM', 'GRU', 'RNN_TANH' or 'RNN_RELU']"""
+                )
+            self.rnn = nn.RNN(ninp,
+                              nhid,
+                              nlayers,
+                              nonlinearity=nonlinearity,
+                              dropout=dropout)
+        self.decoder = nn.Linear(nhid, ntoken)
+
+        # Optionally tie weights as in:
+        # "Using the Output Embedding to Improve Language Models" (Press & Wolf 2016)
+        # https://arxiv.org/abs/1608.05859
+        # and
+        # "Tying Word Vectors and Word Classifiers: A Loss Framework for Language Modeling" (Inan et al. 2016)
+        # https://arxiv.org/abs/1611.01462
+        if tie_weights:
+            if nhid != ninp:
+                raise ValueError(
+                    'When using the tied flag, nhid must be equal to emsize')
+            self.decoder.weight = self.encoder.weight
+
+        self.init_weights()
+
+        self.rnn_type = rnn_type
+        self.nhid = nhid
+        self.nlayers = nlayers
+
+    def init_weights(self):
+        initrange = 0.1
+        nn.init.uniform_(self.encoder.weight, -initrange, initrange)
+        nn.init.zeros_(self.decoder.weight)
+        nn.init.uniform_(self.decoder.weight, -initrange, initrange)
+
+    def forward(self, input, hidden):
+        # import pdb; pdb.set_trace()
```

Comment:
would be nice to have the dimensions commented here, e.g. is it (batch_size, num_steps)?

Reply:
fixed